### PR TITLE
Fix `norm(::HermLocalGenus)` in the non-dyadic ramified case

### PR DESCRIPTION
**Pull request by StevellM**
There was a missing case for the norm computations, namely when the prime ideal was fixed by the involution: in that case, the norm valuation is half the scale valuation.

The added example in test:
```
julia> norm(H2)
1//1 * <5, (z_25 + 1//z_25) + 3>
Norm: 5
Minimum: 5
basis_matrix 
[5 0 0 0 0 0 0 0 0 0; 3 1 0 0 0 0 0 0 0 0; 1 0 1 0 0 0 0 0 0 0; 2 0 0 1 0 0 0 0 0 0; 4 0 0 0 1 0 0 0 0 0; 3 0 0 0 0 1 0 0 0 0; 1 0 0 0 0 0 1 0 0 0; 2 0 0 0 0 0 0 1 0 0; 4 0 0 0 0 0 0 0 1 0; 3 0 0 0 0 0 0 0 0 1]
two normal wrt: 5
```

Before the fix
```
julia> norm(genus(H2))
1//1 * <25, (z_25 + 1//z_25)^2 + 6*(z_25 + 1//z_25) + 9>
Norm: 25
two normal wrt: 5
```
After the fix
```
julia> norm(genus(H2))
1//1 * <5, (z_25 + 1//z_25) + 3>
Minimum: 5
```

### CODE REVIEW

**Comment by StevellM** [comment]
@simonbrandhorst could you have a quick look before merging ?

**Comment by thofma** [comment]
@simonbrandhorst does this look good to you?

File: test/QuadForm/Herm/Genus.jl [hunk 1]
```diff
@@ -420,6 +420,19 @@
...
 
+  # Fix norm in non-dyadic ramified case
+  Qx, x = polynomial_ring(FlintQQ, "x")
+  f = x^10 - 10*x^8 + 35*x^6 + x^5 - 50*x^4 - 5*x^3 + 25*x^2 + 5*x - 1
+  K, a = number_field(f, "a", cached = false)
+  Kt, t = polynomial_ring(K, "t")
+  g = t^2 - a*t + 1
+  E, b = number_field(g, "b", cached = false)
+  D = matrix(E, 1, 1, [1])
+  gens = Vector{Hecke.NfRelElem{nf_elem}}[map(E, [5]), map(E, [a + 3]), map(E, [a^2 + 1]), map(E, [a^3 + 2]), map(E, [a^4 + 4]), map(E, [a^5 + 3]), map(E, [a^6 + 1]), map(E, [a^7 + 2]), map(E, [a^8 + 4]), map(E, [a^9 + 3]), map(E, [b + 4]), map(E, [a*b + 4*a]), map(E, [a^2*b + 4*a^2]), map(E, [a^3*b + 4*a^3]), map(E, [a^4*b + 4*a^4]), map(E, [a^5*b + 4*a^5]), map(E, [a^6*b + 4*a^6]), map(E, [a^7*b + 4*a^7]), map(E, [a^8*b + 4*a^8]), map(E, [a^9*b + 4*a^9])]
```
Comment (thofma) [inline test/QuadForm/Herm/Genus.jl:431]:
```suggestion
  gens = Vector{Hecke.NfRelElem{AbsSimpleNumFieldElem}}[map(E, [5]), map(E, [a + 3]), map(E, [a^2 + 1]), map(E, [a^3 + 2]), map(E, [a^4 + 4]), map(E, [a^5 + 3]), map(E, [a^6 + 1]), map(E, [a^7 + 2]), map(E, [a^8 + 4]), map(E, [a^9 + 3]), map(E, [b + 4]), map(E, [a*b + 4*a]), map(E, [a^2*b + 4*a^2]), map(E, [a^3*b + 4*a^3]), map(E, [a^4*b + 4*a^4]), map(E, [a^5*b + 4*a^5]), map(E, [a^6*b + 4*a^6]), map(E, [a^7*b + 4*a^7]), map(E, [a^8*b + 4*a^8]), map(E, [a^9*b + 4*a^9])]
```

File: src/QuadForm/Herm/Genus.jl [hunk 2]
```diff
@@ -255,8 +255,12 @@ $\mathfrak p$-valuation.
 function norm(G::HermLocalGenus)
   p = prime(G)
   OK = order(p)
-  if !is_dyadic(G) || !is_ramified(G)
+  if !is_ramified(G)
     return fractional_ideal(OK, p)^scale(G, 1)
+  elseif !is_dyadic(G)
+    # If the prime ideal ramifies, then p = P^2, so we need to divide the scale
+    # valuation by 2 to obtain the correct norm valuation
+    return fractional_ideal(OK, p)^(div(scale(G, 1), 2))
```
Comment (simonbrandhorst) [inline src/QuadForm/Herm/Genus.jl:263]: I think that this is wrong. There are subnormal planes in the ramified non-dyadic case. 
Namely the hyperbolic plane should be subnormal. Here the norm is smaller than the scale.

File: test/QuadForm/Herm/Genus.jl [hunk 3]
```diff
@@ -420,6 +420,19 @@
...
+  # Fix norm in non-dyadic ramified case
+  Qx, x = polynomial_ring(FlintQQ, "x")
+  f = x^10 - 10*x^8 + 35*x^6 + x^5 - 50*x^4 - 5*x^3 + 25*x^2 + 5*x - 1
+  K, a = number_field(f, "a", cached = false)
+  Kt, t = polynomial_ring(K, "t")
+  g = t^2 - a*t + 1
+  E, b = number_field(g, "b", cached = false)
+  D = matrix(E, 1, 1, [1])
+  gens = Vector{Hecke.NfRelElem{AbsSimpleNumFieldElem}}[map(E, [5]), map(E, [a + 3]), map(E, [a^2 + 1]), map(E, [a^3 + 2]), map(E, [a^4 + 4]), map(E, [a^5 + 3]), map(E, [a^6 + 1]), map(E, [a^7 + 2]), map(E, [a^8 + 4]), map(E, [a^9 + 3]), map(E, [b + 4]), map(E, [a*b + 4*a]), map(E, [a^2*b + 4*a^2]), map(E, [a^3*b + 4*a^3]), map(E, [a^4*b + 4*a^4]), map(E, [a^5*b + 4*a^5]), map(E, [a^6*b + 4*a^6]), map(E, [a^7*b + 4*a^7]), map(E, [a^8*b + 4*a^8]), map(E, [a^9*b + 4*a^9])]
+  L = hermitian_lattice(E, gens, gram = D)
```
Comment (simonbrandhorst) [inline test/QuadForm/Herm/Genus.jl:432]: I do not understand this example. Can you put a more meaningful test? e.g. a hyperbolic plane  
and a diagonal form.

File: src/QuadForm/Herm/Genus.jl (hunk 2 above)
Comment (StevellM) [reply src/QuadForm/Herm/Genus.jl:263]: I looked at Markus habilitation and yes, I forgot about subnormal sums of planes. This it is a bit more than this: the blocks of even scale valuations are diagonal so we must divide the valuation, and the ones of odd valuation are subnormal and the valuation stays the same. Then we must take the minimum of these new valuations. I will push the fix once the tests pass.

File: test/QuadForm/Herm/Genus.jl (hunk 3 above)
Comment (StevellM) [reply test/QuadForm/Herm/Genus.jl:432]: Yes this was a fancy way for writing a rank 1 lattice with even scale valuation. I will add a hyperbolic plane also.

Comment (simonbrandhorst) [reply test/QuadForm/Herm/Genus.jl]: Is line 431 automatically generated? it seems to me it could be replaced by a single map call.

Comment (StevellM) [reply test/QuadForm/Herm/Genus.jl]: This was the case I was working with when I found the bug, this is why I have added it

Comment (StevellM) [reply test/QuadForm/Herm/Genus.jl]: It is automatically generated yes, via the `Hecke.to_hecke` function.

**Review by simonbrandhorst — APPROVED** [review]

File: test/QuadForm/Herm/Genus.jl [hunk 4]
```diff
@@ -420,6 +420,28 @@
...
+
+  # Rank 1 example
+  Qx, x = polynomial_ring(FlintQQ, "x")
+  f = x^10 - 10*x^8 + 35*x^6 + x^5 - 50*x^4 - 5*x^3 + 25*x^2 + 5*x - 1
+  K, a = number_field(f, "a", cached = false)
+  Kt, t = polynomial_ring(K, "t")
+  g = t^2 - a*t + 1
+  E, b = number_field(g, "b", cached = false)
+  D = matrix(E, 1, 1, [1])
+  gens = Vector{Hecke.NfRelElem{AbsSimpleNumFieldElem}}[map(E, [5]), map(E, [a + 3]), map(E, [a^2 + 1]), map(E, [a^3 + 2]), map(E, [a^4 + 4]), map(E, [a^5 + 3]), map(E, [a^6 + 1]), map(E, [a^7 + 2]), map(E, [a^8 + 4]), map(E, [a^9 + 3]), map(E, [b + 4]), map(E, [a*b + 4*a]), map(E, [a^2*b + 4*a^2]), map(E, [a^3*b + 4*a^3]), map(E, [a^4*b + 4*a^4]), map(E, [a^5*b + 4*a^5]), map(E, [a^6*b + 4*a^6]), map(E, [a^7*b + 4*a^7]), map(E, [a^8*b + 4*a^8]), map(E, [a^9*b + 4*a^9])]
```
Comment (thofma) [inline test/QuadForm/Herm/Genus.jl:433]:
```suggestion
  gens = Vector{RelSimpleNumFieldElem{AbsSimpleNumFieldElem}}[map(E, [5]), map(E, [a + 3]), map(E, [a^2 + 1]), map(E, [a^3 + 2]), map(E, [a^4 + 4]), map(E, [a^5 + 3]), map(E, [a^6 + 1]), map(E, [a^7 + 2]), map(E, [a^8 + 4]), map(E, [a^9 + 3]), map(E, [b + 4]), map(E, [a*b + 4*a]), map(E, [a^2*b + 4*a^2]), map(E, [a^3*b + 4*a^3]), map(E, [a^4*b + 4*a^4]), map(E, [a^5*b + 4*a^5]), map(E, [a^6*b + 4*a^6]), map(E, [a^7*b + 4*a^7]), map(E, [a^8*b + 4*a^8]), map(E, [a^9*b + 4*a^9])]
```

File: test/QuadForm/Herm/Genus.jl [hunk 5]
```diff
@@ -420,6 +420,28 @@
...
+
+  # Rank 1 example
+  Qx, x = polynomial_ring(FlintQQ, "x")
+  f = x^10 - 10*x^8 + 35*x^6 + x^5 - 50*x^4 - 5*x^3 + 25*x^2 + 5*x - 1
+  K, a = number_field(f, "a", cached = false)
+  Kt, t = polynomial_ring(K, "t")
+  g = t^2 - a*t + 1
+  E, b = number_field(g, "b", cached = false)
+  D = matrix(E, 1, 1, [1])
+  gens = Vector{RelSimpleNumFieldElem{AbsSimpleNumFieldElem}}[map(E, [5]), map(E, [a + 3]), map(E, [a^2 + 1]), map(E, [a^3 + 2]), map(E, [a^4 + 4]), map(E, [a^5 + 3]), map(E, [a^6 + 1]), map(E, [a^7 + 2]), map(E, [a^8 + 4]), map(E, [a^9 + 3]), map(E, [b + 4]), map(E, [a*b + 4*a]), map(E, [a^2*b + 4*a^2]), map(E, [a^3*b + 4*a^3]), map(E, [a^4*b + 4*a^4]), map(E, [a^5*b + 4*a^5]), map(E, [a^6*b + 4*a^6]), map(E, [a^7*b + 4*a^7]), map(E, [a^8*b + 4*a^8]), map(E, [a^9*b + 4*a^9])]
```
Comment (thofma) [inline test/QuadForm/Herm/Genus.jl:433]:
```suggestion
  gens = Vector{Hecke.RelSimpleNumFieldElem{AbsSimpleNumFieldElem}}[map(E, [5]), map(E, [a + 3]), map(E, [a^2 + 1]), map(E, [a^3 + 2]), map(E, [a^4 + 4]), map(E, [a^5 + 3]), map(E, [a^6 + 1]), map(E, [a^7 + 2]), map(E, [a^8 + 4]), map(E, [a^9 + 3]), map(E, [b + 4]), map(E, [a*b + 4*a]), map(E, [a^2*b + 4*a^2]), map(E, [a^3*b + 4*a^3]), map(E, [a^4*b + 4*a^4]), map(E, [a^5*b + 4*a^5]), map(E, [a^6*b + 4*a^6]), map(E, [a^7*b + 4*a^7]), map(E, [a^8*b + 4*a^8]), map(E, [a^9*b + 4*a^9])]
```

**Pull request merged**